### PR TITLE
Head-to-head rapidjson benchmarks

### PR DESCRIPTION
**Pull request by jkeiser**
This adds usage-based tests (vs. ondemand) of rapidjson to our benchmarks.

### Skylake (clang10)

```
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                          Time             CPU   Iterations best_branch_miss best_bytes_per_sec best_cache_miss best_cache_ref best_cycles best_cycles_per_byte best_docs_per_sec best_frequency best_instructions best_instructions_per_byte best_instructions_per_cycle best_items_per_sec branch_miss      bytes bytes_per_second cache_miss  cache_ref     cycles cycles_per_byte docs_per_sec  frequency instructions instructions_per_byte instructions_per_cycle      items items_per_second
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
partial_tweets<simdjson_dom>                  273858 ns       273861 ns         2556           3.246k           2.39074G               0        87.723k    975.314k               1.5444          3.78573k       3.69227G          2.99617M                    4.74442                     3.07201           378.573k    3.39685k   631.515k        2.1476G/s  0.0974178   87.6076k   980.674k         1.55289   3.65149k/s 3.58093G/s     2.99617M               4.74442                3.05521        100       365.149k/s [BEST: throughput=  2.39 GB/s doc_throughput=  3785 docs/s instructions=     2996170 cycles=      975314 branch_miss=    3246 cache_miss=       0 cache_ref=     87723 items=       100 avg_time=    265741 ns]
partial_tweets<simdjson_ondemand>             171775 ns       171778 ns         4131           1.878k           3.88522G               0        54.998k      600.3k             0.950571          6.15222k       3.69318G          1.94802M                    3.08467                     3.24507           615.222k    2.02482k   631.515k       3.42387G/s  0.0317115   54.9065k   604.907k        0.957866   5.82148k/s 3.52145G/s     1.94802M               3.08467                3.22036        100       582.148k/s [BEST: throughput=  3.89 GB/s doc_throughput=  6152 docs/s instructions=     1948018 cycles=      600300 branch_miss=    1878 cache_miss=       0 cache_ref=     54998 items=       100 avg_time=    163870 ns]
partial_tweets<yyjson>                        637952 ns       637950 ns         1102           9.981k            1.5296G             223          69.4k    1.35699M              2.14879          2.42212k        3.2868G           2.9024M                    4.59593                     2.13885           242.212k    7.69653k   631.515k       944.054M/s    944.392   49.3342k   1.52222M         2.41042   1.56752k/s 2.38611G/s     2.90284M               4.59662                1.90698        100       156.752k/s [BEST: throughput=  1.53 GB/s doc_throughput=  2422 docs/s instructions=     2902398 cycles=     1356993 branch_miss=    9981 cache_miss=     223 cache_ref=     69400 items=       100 avg_time=    629182 ns]
partial_tweets<rapidjson>                    2379573 ns      2379615 ns          291          28.951k           289.494M         15.168k        52.256k    8.05268M              12.7514           458.412       3.69145G          21.8987M                    34.6765                     2.71943           45.8412k    29.3319k   631.515k       253.091M/s    618.873   32.1571k   8.18529M         12.9614    420.236/s 3.43975G/s     21.8994M               34.6776                2.67546        100       42.0236k/s [BEST: throughput=  0.29 GB/s doc_throughput=   458 docs/s instructions=    21898743 cycles=     8052684 branch_miss=   28951 cache_miss=   15168 cache_ref=     52256 items=       100 avg_time=   2371117 ns]

large_random<simdjson_dom>                  88975940 ns     88977383 ns            8         933.553k           517.703M        10.9341M       15.2251M    327.801M              7.12629           11.2547       3.68931G          1039.39M                    22.5961                     3.17081           11.2547M    932.292k   45.9988M       493.022M/s   10.9997M   15.2317M   328.201M           7.135    11.2388/s 3.68859G/s     1039.39M               22.5961                3.16694      1000k       11.2388M/s [BEST: throughput=  0.52 GB/s doc_throughput=    11 docs/s instructions=  1039392795 cycles=   327800845 branch_miss=  933553 cache_miss=10934061 cache_ref=  15225147 items=   1000000 avg_time=  88963075 ns]
large_random<simdjson_ondemand>             66949764 ns     66950838 ns           10         913.217k           687.527M         5.7062M       7.90104M    246.836M              5.36614           14.9466       3.68936G          684.848M                    14.8884                     2.77451           14.9466M    913.408k   45.9988M       655.225M/s   5.73947M    7.9028M   246.939M         5.36838    14.9363/s 3.68836G/s     684.848M               14.8884                2.77335      1000k       14.9363M/s [BEST: throughput=  0.69 GB/s doc_throughput=    14 docs/s instructions=   684847924 cycles=   246835832 branch_miss=  913217 cache_miss= 5706203 cache_ref=   7901045 items=   1000000 avg_time=  66937458 ns]
large_random<simdjson_ondemand_unordered>   66942526 ns     66943605 ns           10         905.074k           687.607M        5.68882M       7.89594M    246.803M              5.36542           14.9484        3.6893G          690.848M                    15.0188                     2.79919           14.9484M    905.144k   45.9988M       655.296M/s    5.7195M   7.89743M   246.921M         5.36799    14.9379/s 3.68849G/s     690.848M               15.0188                2.79785      1000k       14.9379M/s [BEST: throughput=  0.69 GB/s doc_throughput=    14 docs/s instructions=   690847916 cycles=   246802730 branch_miss=  905074 cache_miss= 5688821 cache_ref=   7895937 items=   1000000 avg_time=  66930326 ns]
large_random<yyjson>                       121352101 ns    121354070 ns            6         661.944k           379.277M        5.87944M       9.04773M     330.53M              7.18562           8.24538       2.72534G           866.68M                    18.8414                     2.62209           8.24538M    662.095k   45.9988M       361.486M/s   5.87906M   9.07484M   330.597M         7.18708    8.24035/s 2.72423G/s      866.68M               18.8414                2.62156      1000k       8.24035M/s [BEST: throughput=  0.38 GB/s doc_throughput=     8 docs/s instructions=   866680092 cycles=   330529919 branch_miss=  661944 cache_miss= 5879437 cache_ref=   9047731 items=   1000000 avg_time= 121334309 ns]
large_random<rapidjson>                    212464105 ns    212467661 ns            3         594.462k           222.299M        6.97403M       10.6864M    707.076M              15.3716           4.83272        3.4171G          2.01615G                    43.8305                     2.85139           4.83272M    589.684k   45.9988M       206.468M/s   6.50889M   10.0156M    709.28M         15.4195     4.7066/s  3.3383G/s     2.01616G               43.8307                2.84254      1000k        4.7066M/s [BEST: throughput=  0.22 GB/s doc_throughput=     4 docs/s instructions=  2016147826 cycles=   707075575 branch_miss=  594462 cache_miss= 6974029 cache_ref=  10686411 items=   1000000 avg_time= 212444721 ns]
large_random<rapidjson_lossless>           244392463 ns    244396510 ns            3         582.208k           192.606M        6.97342M       10.6878M    824.748M              17.9298           4.18719       3.45338G          2.47511G                    53.8083                     3.00106           4.18719M    582.576k   45.9988M       179.495M/s   6.50319M   10.0197M   827.032M         17.9794    4.09171/s 3.38398G/s     2.47512G               53.8085                2.99278      1000k       4.09171M/s [BEST: throughput=  0.19 GB/s doc_throughput=     4 docs/s instructions=  2475114932 cycles=   824748223 branch_miss=  582208 cache_miss= 6973415 cache_ref=  10687810 items=   1000000 avg_time= 244372758 ns]
large_random<rapidjson_insitu>             259302564 ns    259306851 ns            3         597.688k           181.228M        8.96115M       13.5822M    842.756M              18.3213           3.93984       3.32032G          1.74727G                    37.9851                     2.07328           3.93984M    597.084k   45.9988M       169.174M/s   8.52804M   12.9162M   844.864M         18.3671    3.85643/s 3.25816G/s     1.74728G               37.9853                2.06812      1000k       3.85643M/s [BEST: throughput=  0.18 GB/s doc_throughput=     3 docs/s instructions=  1747268570 cycles=   842755793 branch_miss=  597688 cache_miss= 8961147 cache_ref=  13582200 items=   1000000 avg_time= 259283292 ns]

kostya<simdjson_dom>                        92538176 ns     92539692 ns            8         1027.55k            1.4847G        15.7762M       22.2641M    341.173M              2.48478           10.8131       3.68914G          985.296M                    7.17596                     2.88797           5.66919M    1027.94k   137.305M       1.38184G/s   15.8048M   22.1771M   341.339M         2.48599    10.8062/s 3.68857G/s     985.296M               7.17596                2.88656   524.288k       5.66555M/s [BEST: throughput=  1.48 GB/s doc_throughput=    10 docs/s instructions=   985295809 cycles=   341172695 branch_miss= 1027549 cache_miss=15776247 cache_ref=  22264119 items=    524288 avg_time=  92524925 ns]
kostya<simdjson_ondemand>                   63952980 ns     63954002 ns           11          465.79k           2.14937G        9.99833M       13.9283M    235.671M              1.71641            15.654        3.6892G          658.901M                    4.79882                     2.79585           8.20721M    466.127k   137.305M       1.99949G/s   10.0382M   13.9272M   235.878M         1.71791    15.6362/s 3.68824G/s     658.901M               4.79882                 2.7934   524.288k       8.19789M/s [BEST: throughput=  2.15 GB/s doc_throughput=    15 docs/s instructions=   658901372 cycles=   235671225 branch_miss=  465790 cache_miss= 9998335 cache_ref=  13928284 items=    524288 avg_time=  63940562 ns]
kostya<yyjson>                             175642462 ns    175645361 ns            4         419.052k           782.028M        11.3416M       16.2911M    453.885M              3.30567           5.69555       2.58513G          978.903M                     7.1294                     2.15672           2.98611M    417.705k   137.305M       745.504M/s   11.3439M   16.3195M   454.141M         3.30753    5.69329/s 2.58556G/s     978.903M                7.1294                 2.1555   524.288k       2.98492M/s [BEST: throughput=  0.78 GB/s doc_throughput=     5 docs/s instructions=   978902663 cycles=   453885225 branch_miss=  419052 cache_miss=11341571 cache_ref=  16291122 items=    524288 avg_time= 175623725 ns]
kostya<rapidjson>                          274726100 ns    274726027 ns            3         946.542k           505.748M         8.3209M       12.2058M    921.307M              6.70993           3.68339       3.39354G           2.7424G                     19.973                     2.97664           1.93116M     947.13k   137.305M       476.636M/s   8.02636M   11.7596M   923.755M         6.72776    3.63999/s 3.36246G/s      2.7424G               19.9731                2.96875   524.288k        1.9084M/s [BEST: throughput=  0.51 GB/s doc_throughput=     3 docs/s instructions=  2742397940 cycles=   921307244 branch_miss=  946542 cache_miss= 8320900 cache_ref=  12205797 items=    524288 avg_time= 274707614 ns]
kostya<rapidjson_lossless>                 301750804 ns    301755744 ns            2         960.446k           458.773M         8.3266M       12.2055M    1023.81M              7.45646           3.34127       3.42083G          3.00768G                    21.9051                     2.93773           1.75179M    959.167k   137.305M       433.941M/s   8.09626M   11.8656M   1025.79M         7.47091    3.31394/s 3.39942G/s     3.00768G               21.9051                2.93205   524.288k       1.73746M/s [BEST: throughput=  0.46 GB/s doc_throughput=     3 docs/s instructions=  3007678363 cycles=  1023809721 branch_miss=  960446 cache_miss= 8326604 cache_ref=  12205533 items=    524288 avg_time= 301730438 ns]
kostya<rapidjson_insitu>                   330829370 ns    330824560 ns            2         1001.08k           417.968M        14.6336M       19.9756M    1020.41M              7.43172           3.04408       3.10622G          2.23976G                    16.3123                     2.19496           1.59598M    1000.03k   137.305M       395.812M/s   14.4012M   19.7411M   1021.87M         7.44235    3.02275/s 3.08886G/s     2.23976G               16.3123                2.19182   524.288k       1.58479M/s [BEST: throughput=  0.42 GB/s doc_throughput=     3 docs/s instructions=  2239762299 cycles=  1020412119 branch_miss= 1001082 cache_miss=14633556 cache_ref=  19975576 items=    524288 avg_time= 330808790 ns]

distinct_user_id<simdjson_dom>                269610 ns       269615 ns         2595           3.424k           2.44605G               0        88.465k    953.228k              1.50943          3.87331k       3.69215G          2.93045M                    4.64034                     3.07423           445.431k    3.59464k   631.515k       2.18143G/s  0.0416185   88.5125k   958.458k         1.51771     3.709k/s 3.55492G/s     2.93045M               4.64034                3.05746        115       426.535k/s [BEST: throughput=  2.45 GB/s doc_throughput=  3873 docs/s instructions=     2930445 cycles=      953228 branch_miss=    3424 cache_miss=       0 cache_ref=     88465 items=       115 avg_time=    259723 ns]
distinct_user_id<simdjson_ondemand>           155274 ns       155276 ns         4508           1.376k           4.35798G               0        52.589k    535.171k              0.84744          6.90084k       3.69313G          1.89752M                    3.00471                     3.54563           793.596k     1.5022k   631.515k       3.78773G/s   0.011535   52.6373k   538.782k        0.853157   6.44014k/s 3.46983G/s     1.89752M               3.00471                3.52187        115       740.616k/s [BEST: throughput=  4.36 GB/s doc_throughput=  6900 docs/s instructions=     1897519 cycles=      535171 branch_miss=    1376 cache_miss=       0 cache_ref=     52589 items=       115 avg_time=    145975 ns]
distinct_user_id<yyjson>                      638583 ns       638594 ns         1198          10.043k           1.54716G             149        66.236k    1.33898M              2.12026          2.44992k        3.2804G          2.86143M                    4.53106                     2.13702           281.741k    7.81875k   631.515k       943.102M/s    992.891   48.1001k   1.51146M         2.39339   1.56594k/s 2.36685G/s     2.86178M               4.53162                1.89339        115       180.083k/s [BEST: throughput=  1.55 GB/s doc_throughput=  2449 docs/s instructions=     2861430 cycles=     1338979 branch_miss=   10043 cache_miss=     149 cache_ref=     66236 items=       115 avg_time=    626150 ns]
distinct_user_id<rapidjson>                  2359660 ns      2359702 ns          296          28.974k           290.619M         14.683k        53.939k    8.02162M              12.7022           460.193       3.69149G          21.8276M                    34.5638                     2.72109           52.9222k    29.2952k   631.515k       255.227M/s   2.69226k   35.1304k   8.15739M         12.9172    423.782/s 3.45696G/s     21.8284M               34.5651                 2.6759        115        48.735k/s [BEST: throughput=  0.29 GB/s doc_throughput=   460 docs/s instructions=    21827568 cycles=     8021616 branch_miss=   28974 cache_miss=   14683 cache_ref=     53939 items=       115 avg_time=   2349585 ns]
distinct_user_id<rapidjson_insitu>           1558554 ns      1558532 ns          436          25.423k           437.429M          7.086k        73.026k    5.32941M              8.43909           692.666        3.6915G          13.2491M                    20.9799                     2.48604           79.6565k    28.9602k   631.515k       386.428M/s   3.01785k   64.7873k   5.45254M         8.63407    641.629/s 3.49851G/s     13.2495M               20.9804                2.42996        115       73.7874k/s [BEST: throughput=  0.44 GB/s doc_throughput=   692 docs/s instructions=    13249139 cycles=     5329413 branch_miss=   25423 cache_miss=    7086 cache_ref=     73026 items=       115 avg_time=   1548647 ns]

find_tweet<simdjson_dom>                      256383 ns       256387 ns         2734           3.103k           2.55808G               0        78.441k    911.548k              1.44343           4.0507k       3.69241G          2.84527M                    4.50547                     3.12136            4.0507k     3.2419k   631.515k       2.29397G/s    1.20922   78.3838k   916.295k         1.45095   3.90035k/s 3.57387G/s     2.84527M               4.50547                3.10519          1       3.90035k/s [BEST: throughput=  2.56 GB/s doc_throughput=  4050 docs/s instructions=     2845271 cycles=      911548 branch_miss=    3103 cache_miss=       0 cache_ref=     78441 items=         1 avg_time=    248292 ns]
find_tweet<simdjson_ondemand>                 116312 ns       116314 ns         6015              673           5.85159G               0        27.212k    398.665k             0.631284          9.26595k       3.69401G          1.39034M                     2.2016                     3.48749           9.26595k     719.171   631.515k        5.0565G/s   0.108396    27.243k   400.373k        0.633988   8.59738k/s 3.44216G/s     1.39034M                2.2016                3.47261          1       8.59738k/s [BEST: throughput=  5.85 GB/s doc_throughput=  9265 docs/s instructions=     1390341 cycles=      398665 branch_miss=     673 cache_miss=       0 cache_ref=     27212 items=         1 avg_time=    108449 ns]
find_tweet<yyjson>                            625160 ns       625171 ns         1220          10.002k           1.57896G             128        56.991k    1.30995M              2.07431          2.50028k       3.27525G          2.80074M                    4.43496                     2.13805           2.50028k    7.51238k   631.515k       963.352M/s    1081.16   39.6875k   1.47578M         2.33689   1.59956k/s  2.3606G/s     2.80117M               4.43564                 1.8981          1       1.59956k/s [BEST: throughput=  1.58 GB/s doc_throughput=  2500 docs/s instructions=     2800743 cycles=     1309955 branch_miss=   10002 cache_miss=     128 cache_ref=     56991 items=         1 avg_time=    616330 ns]
find_tweet<rapidjson>                        2347925 ns      2347876 ns          298          28.707k           291.564M         14.781k        46.306k    7.99556M              12.6609            461.69       3.69147G          21.7433M                    34.4304                     2.71943             461.69    29.1269k   631.515k       256.513M/s   2.37769k   27.3119k    8.1185M         12.8556    425.917/s 3.45781G/s     21.7442M               34.4318                2.67835          1        425.917/s [BEST: throughput=  0.29 GB/s doc_throughput=   461 docs/s instructions=    21743329 cycles=     7995561 branch_miss=   28707 cache_miss=   14781 cache_ref=     46306 items=         1 avg_time=   2339574 ns]
find_tweet<rapidjson_insitu>                 1554890 ns      1554917 ns          441          25.228k           441.462M           5.81k        63.152k    5.28065M              8.36187           699.053       3.69145G          13.1707M                    20.8558                     2.49415            699.053    30.3394k   631.515k       387.326M/s   2.82039k    55.271k   5.44445M         8.62125    643.121/s 3.50144G/s      13.171M               20.8562                2.41917          1        643.121/s [BEST: throughput=  0.44 GB/s doc_throughput=   699 docs/s instructions=    13170720 cycles=     5280649 branch_miss=   25228 cache_miss=    5810 cache_ref=     63152 items=         1 avg_time=   1546667 ns]
```

### CODE REVIEW

**Comment by jkeiser** [comment]
I didn't shoehorn insitu yet ... will take a bit more work (because we need to copy the doc every time again, but I'd rather not include that in the benchmark time).

**Comment by jkeiser** [comment]
@lemire which implementations do you recommend focusing on in the modern age? My plan was to focus on competition-core (nlohmann_json rapidjson sajson cjson jsmn yyjson boostjson) and not do competition-all (jsoncpp json11 fastjson gason ujson4c) ... but if I can do less and get most of the mileage, all the better :)

**Comment by jkeiser** [comment]
OK, insitu is in. The cost of the memcpy is added to the cost of the test; there are ways to make that not be so, but I'm not planning to mess with them until I hit an insitu implementation that's a little closer in performance (where it might matter more).

**Comment by lemire** [comment]
> which implementations do you recommend focusing on in the modern age? My plan was to focus on competition-core (nlohmann_json rapidjson sajson cjson jsmn yyjson boostjson) and not do competition-all (jsoncpp json11 fastjson gason ujson4c) ... but if I can do less and get most of the mileage, all the better :)

I agree. There are diminishing returns to extending the scope beyond a certain point. Where you cut is a judgement call, of course...

File: benchmark/distinct_user_id/yyjson.h [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 namespace distinct_user_id {
 
-class yyjson {
+struct yyjson {
 public:
```
Comment (lemire) [inline benchmark/distinct_user_id/yyjson.h:10]: If you switch to a struct, you may not need the ugly public qualifier.

File: benchmark/find_tweet/rapidjson.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  bool run(Document &root, uint64_t find_id, std::string_view &text) {
+    if (root.HasParseError() || !root.IsObject()) { return false; }
+    auto statuses = root.FindMember("statuses");
+    if (statuses == root.MemberEnd() || !statuses->value.IsArray()) { return false; }
+    for (auto &tweet : statuses->value.GetArray()) {
+      if (!tweet.IsObject()) { return false; }
+      auto id = tweet.FindMember("id");
+      if (id == tweet.MemberEnd() || !id->value.IsUint64()) { return false; }
+      if (id->value.GetUint64() == find_id) {
+        auto _text = tweet.FindMember("text");
```
Comment (lemire) [inline benchmark/find_tweet/rapidjson.h:23]: I am fine with `_text` as a variable name but some people might object.

File: benchmark/partial_tweets/rapidjson.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+};
+
+struct rapidjson : public rapidjson_base {
+  bool run(const padded_string &json, std::vector<tweet> &tweets) {
+    return rapidjson_base::run(doc.Parse<kParseValidateEncodingFlag>(json.data()), tweets);
+  }
+};
+BENCHMARK_TEMPLATE(partial_tweets, rapidjson);
+
+// TODO this fails!
```
Comment (lemire) [inline benchmark/partial_tweets/rapidjson.h:69]: Why does adding a copy fail?

**Review by lemire — APPROVED** [review]
Very good. Of course, once this is done, then it calls for removing my old benchmarking code as it becomes squarely obsolete. This is much better!!!

File: benchmark/partial_tweets/rapidjson.h (hunk 3 above)
Comment (jkeiser) [reply benchmark/partial_tweets/rapidjson.h:69]: It's not the copy,  it's the in situ parsing.  The resulting string doesn't get unescaped in this case,  for some reason. I verified by switching from ParseInsitu to Parse, and it succeeded. 

I'll file with rapidjson.

File: benchmark/distinct_user_id/yyjson.h (hunk 1 above)
Comment (jkeiser) [reply benchmark/distinct_user_id/yyjson.h:10]: That was the purpose :) Looks like I missed that one, thanks!

File: benchmark/find_tweet/rapidjson.h (hunk 2 above)
Comment (jkeiser) [reply benchmark/find_tweet/rapidjson.h:23]: Yeah ... I've been thinking about renaming all the result values "result," which would get rid of the need for things like this and make more of the code copy/pasteable. Probably do that when I implement sajson (the last one, I think).

File: benchmark/partial_tweets/rapidjson.h (hunk 3 above)
Comment (jkeiser) [reply benchmark/partial_tweets/rapidjson.h:69]: (The find_tweet code works fine insitu, fyi--it's just this code that fails to unescape. Not sure why.)

**Pull request merged**